File: General-DSA-Questions/Some Named Algo & Genral Ques/8_binarySearchInRotatedArray.cpp
// Write a function for binary search in a rotated array.

// You are given a sorted array that has been rotated at an unknown pivot.
// Given a target value, return its index if found; otherwise return -1.

// Input: nums = [4,5,6,7,0,1,2], target = 0  
// Output: 4  

// Video link : https://www.youtube.com/watch?v=6WNZQBHWQJs&list=PLfqMhTWNBTe137I_EPQd34TsgV6IO55pt&index=18

// 🧠 Intuition:
// Even though the array is rotated, at least one half (left or right) of the array is always sorted.
// We can use that fact to decide where to search.

// How this works :
// 1. If it is a rotated array, then one half of the array must be sorted.
// 2. If the left half is sorted, then check if the target lies in the left half. If it does, then search in the left half, else search in the right half.
//    a. Assuming target lies in left half : Searching in left half is just like normal binary search.
// 3. If the right half is sorted, then check if the target lies in the right half. If it does, then search in the right half, else search in the left half.
//    a. Assuming target lies in right half : Searching in right half is just like normal binary search.

#include <iostream>
#include <vector>

using namespace std;

int binarySearchInRotatedArray(vector<int> &num, int target) {
    int left = 0;   // Left pointer
    int right = num.size() - 1; // Right pointer
    while (left <= right) {
        int mid = left + ((right - left) / 2);
        if (num[mid] == target) {   // If the middle element is the target, return its index
            return mid;
        } else if (num[left] <= num[mid]) { // If the left half is sorted
            if (num[left] <= target && target <= num[mid]) {    // If target is in the left half
                // Search in the left half
                right = mid - 1;
            } else {    // If target is not in the left half, search in the right half
                // Search in the right half
                left = mid + 1;
            }
        } else {    // If the right half is sorted
            if (num[mid] < target && target <= num[right]) {    // If target is in the right half
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
    }
    return -1;
}

int main() {
    vector<int> num = {4, 5, 6, 7, 0, 1, 2};
    int target = 0;
    cout << binarySearchInRotatedArray(num, target) << endl;
    return 0;
}


// There is one more question that is similar to this one, which is to search in a rotated sorted array with duplicates.
// The logic is the same, but you have to handle one edge case where there are duplicates in the array and left, mid, and right elements are equal.
// We cannot determine which side is sorted. In that case, you can just increment the left pointer and decrement the right pointer to avoid infinite loop.
// Example:
// Input: nums = [2, 2, 2, 3, 4, 2], target = 3
// Output: 3

// Notes : https://takeuforward.org/arrays/search-element-in-rotated-sorted-array-ii/

// How to handle duplicates in rotated sorted array:
// 1. If left, mid, and right elements are equal, increment left and decrement right.
// 2. If the left half is sorted, check if the target lies in the left half. If it does, search in the left half; otherwise, search in the right half.
// 3. If the right half is sorted, check if the target lies in the right half. If it does, search in the right half; otherwise, search in the left half.

// //Edge case: Add this condition after checking if target == num[mid] and before checking if left half is sorted.
// if (arr[low] == arr[mid] && arr[mid] == arr[high]) {
//     low = low + 1;   // Increment low to avoid infinite loop
//     high = high - 1; // Decrement high to avoid infinite loop
//     continue;
// }

File: JS-DSA-Questions/Some Named Algo & Genral Ques/8_binarySearchInRotatedArray.js
// Write a function for binary search in a rotated array.

// You are given a sorted array rotated at an unknown pivot. Given a target, return its index or -1.

// Input: nums = [4,5,6,7,0,1,2], target = 0 → Output: 4

// Video link : https://www.youtube.com/watch?v=6WNZQBHWQJs&list=PLfqMhTWNBTe137I_EPQd34TsgV6IO55pt&index=18

// 🧠 Intuition:
// Even though the array is rotated, at least one half (left or right) is always sorted.
// Use that fact to decide where to search.

// How this works:
// 1. In a rotated array, one half must be sorted.
// 2. If the left half is sorted, check whether the target lies within it; search accordingly.
// 3. If the right half is sorted, check whether the target lies within it; search accordingly.

function binarySearchInRotatedArray(num, target) {
  let left = 0;
  let right = num.length - 1;
  while (left <= right) {
    const mid = left + Math.floor((right - left) / 2);
    if (num[mid] === target) {
      return mid;
    } else if (num[left] <= num[mid]) {
      // Left half is sorted
      if (num[left] <= target && target <= num[mid]) {
        right = mid - 1; // target is in the left half
      } else {
        left = mid + 1; // search the right half
      }
    } else {
      // Right half is sorted
      if (num[mid] < target && target <= num[right]) {
        left = mid + 1; // target is in the right half
      } else {
        right = mid - 1; // search the left half
      }
    }
  }
  return -1;
}

console.log(binarySearchInRotatedArray([4, 5, 6, 7, 0, 1, 2], 0)); // 4

// Variant: Search in a rotated sorted array WITH duplicates.
// Same logic, but when num[left] === num[mid] === num[right] we can't tell which side is sorted,
// so we shrink both ends (left++, right--) to avoid an infinite loop:
//
// if (num[left] === num[mid] && num[mid] === num[right]) {
//   left++;
//   right--;
//   continue;
// }
//
// Notes: https://takeuforward.org/arrays/search-element-in-rotated-sorted-array-ii/
